src/odb/src/defout/defout_impl.cpp

```diff
@@ -913,6 +928,69 @@ void defout_impl::writeGroups(dbBlock* block)
   fprintf(_out, "END GROUPS\n");
 }
 
+void defout_impl::writeScanChains(dbBlock* block)
+{
+  dbDft* dft = block->getDft();
+  dbSet<dbScanChain> scan_chains = dft->getScanChains();
+  if (scan_chains.empty()) {
+    // If we don't have scan chains we have nothing to print
+    return;
+  }
+  fprintf(_out, "\nSCANCHAINS %d ;\n\n", scan_chains.size());
+
+  for (dbScanChain* scan_chain : dft->getScanChains()) {
+    dbSet<dbScanPartition> scan_partitions = scan_chain->getScanPartitions();
+    int chain_suffix = 0;
+    for (dbScanPartition* scan_partition : scan_partitions) {
+      bool already_printed_floating = false;
+      bool already_printed_ordered = false;
+      const std::string chain_name
+          = scan_partitions.size() == 1
+                ? scan_chain->getName()
+                : fmt::format("{}_{}", scan_chain->getName(), chain_suffix);
+
+      const std::string start_pin_name = std::visit(
+          [](auto&& pin) { return pin->getName(); }, scan_chain->getScanIn());
+      const std::string stop_pin_name = std::visit(
+          [](auto&& pin) { return pin->getName(); }, scan_chain->getScanOut());
+
+      fprintf(_out, "- %s\n", chain_name.c_str());
+      fprintf(_out, "+ START PIN %s\n", start_pin_name.c_str());
+
+      for (dbScanList* scan_list : scan_partition->getScanLists()) {
+        dbSet<dbScanInst> scan_insts = scan_list->getScanInsts();
+        if (scan_insts.size() == 1 && !already_printed_floating) {
+          fprintf(_out, "+ FLOATING\n");
+          already_printed_floating = true;
+          already_printed_ordered = false;
+        } else if (scan_insts.size() > 1 && !already_printed_ordered) {
+          fprintf(_out, "+ ORDERED\n");
+          already_printed_floating = false;
+          already_printed_ordered = true;
+        }
+
+        for (dbScanInst* scan_inst : scan_insts) {
+          dbScanInst::AccessPins access_pins = scan_inst->getAccessPins();
+          const std::string scan_in_name = std::visit(
+              [](auto&& pin) { return getPinName(pin); }, access_pins.scan_in);
+          const std::string scan_out_name = std::visit(
+              [](auto&& pin) { return getPinName(pin); }, access_pins.scan_out);
+          fprintf(_out,
+                  "  %s ( IN %s ) ( OUT %s )\n",
+                  scan_inst->getInst()->getName().c_str(),
+                  scan_in_name.c_str(),
+                  scan_out_name.c_str());
```

Comment:
Is this correct if you have a bterm for scan_in/out ?

Reply:
Hey Matt,

I am having issues coming up with a test case for this.

I have in verilog something like:

```
module top();
   sub sub1(scan_in_1, scan_out_1);
   sub sub2(scan_in_2, scan_out_2);
endmodule
```


But looks like odb flattens the modules and the instances take names like "sub1/cell". The only way that I found that may work is getting the dbModInst at the top, then get the dbModule master and then get a dbModBTerm:

```
for (dbModInst* inst: block->getModInsts()) {
    if (inst->getName() == std::string_view(inst_name)) {
        dbModule* master = inst->getMaster();
        for (dbModBTerm* bterm:  master->getModBTerms()) {
            std::cout << "GARAY " << bterm->getName() << std::endl; // This gives me scan_in_1 and scan_out_1
        }
    }
}
```

Q1) What is the relationship between dbModBTerm and a dbBTerm? I don't think I can get a BTerm from a ModBTerm.

Q2) Do I need to go back to my dbScanPin and change the BTerm for a dbModBTerm?


Thanks!

Reply:
dbModBTerm is for terms of a dbModule.   Such a term may just represent an intermediate level of the hierarchy with no corresponding dbBTerm.

Are scan pins ever intermediate hierarchy levels?

Reply:
So, I am not sure about that concept in odb so let me try to explain what I want here:

I want to be able to represent a CORE where we don't want to expose the internals in the scandef:

![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/1619948/e6839caf-2d5f-4ac0-b43e-ec243070e3d0)

So, this CORE will have two scan ins and two scan outs. For this core, we have this scandef:

![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/1619948/d9d9919f-5669-402e-ad73-abf5e6dc6a84)

Note that test_si1,  test_so1, test_si2 and test_so2 are pins (are they ITerms or BTerms?).

So now my question would be: how can OpenROAD represent this? Is this going to be a hier? if this is a hier, how can I access those pins?

Thanks for your help Matt!

Reply:
DEF can only refer to instances that appear in the DEF afaik.  That would imply that CORE is a module with a LEF and these are iterms.  Does that seem right?

Reply:
Thanks!, I created a regresion for the cores.

Responding at your first question, I think I got the Bterm/ITerm relationship wrong:

Looking at this: 
![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/1619948/22952cad-38ba-4712-8ffa-fdf176b67c5c)

I think that I will not have a BTerm ever for the cells/blocks and I will always have an ITerm. In that case, the scaninsts do not need to hold a BTerm. Wdyt?

I still think I need that std::variant for BTerm ITerm for scan chains.

Reply:
Can a scan chain start/end on a primary (ie top-level) bterm?  If not then you will always have iterms.

Reply:
Yes!, they can. I think then it makes sense to have the std::variant. Thanks!

Reply:
Ok but bterms don't belong to an instance so is this format correct?